### PR TITLE
Fatal error / exception handling

### PR DESCRIPTION
**Pull request by mikee47**
The goal of this PR is to be really horrible to jerryscript and do nasty things in javascript to see what breaks.

First thing is to allocate large array and see what happens.
We get a fatal 'out of memory' exception which trips 'jerry_port_fatal`.
As this is a no-return function, I've used setjmp/longjmp to catch these failures in `call`, which works.

BUT. How to indicate this to the caller? In general, we cannot rely on the state of jerryscript after a fatal error so it must be re-initialised. That requires application co-operation.

The solution proposed here implements a simple exception handling mechanism using setjmp/longjmp.
Two macros, `JS_TRY()` and `JS_CATCH()` could be re-implemented using regular C++ exception handling if required.

The handler has to go at the top execution level. I've revised the `Event_Jsvm` sample to demonstrate how this works.

Note: Enabling jerryscript assertions makes recovery impossible in some situations, so clearly that's only intended for debugging jerryscript itself.

### CODE REVIEW

**Comment by mikee47** [comment]
@slaff Tried a few approaches to this and think this is the way to go. Thoughts?

**Comment by slaff** [comment]
> Tried a few approaches to this and think this is the way to go. Thoughts?

@mikee47 I like this PR the way it is. If you have some additional work to polish this PR - can you do it so that I can merge this PR and have these changes also in Sming v4.5.0 ?

File: samples/Event_Jsvm/app/application.cpp [hunk 1]
```diff
@@ -67,21 +69,38 @@ bool triggerEvent(const String& name, const JS::Object& params)
 		return false;
 	}
 
-	// Build the event object...
-	JS::Object event;
-	event["name"] = name;
-	event["params"] = params;
+	JS_TRY()
```
Comment (slaff) [inline samples/Event_Jsvm/app/application.cpp:72]: @mikee47 I am wondering isn't it possible to use try-catch directly in the JavaScript code and catch out-of-memory exception? We could expose a `reset()` JavaScript function that resets the VM and give the JS developers (end user) better control. On the C/C++ side we could check if the VM is in error state and restart the VM automatically if it stays in this error state for 8 seconds.

Comment (mikee47) [reply samples/Event_Jsvm/app/application.cpp:72]: Unfortunately fatal errors are not the same as javascript exceptions and cannot be caught in script. The engine is essentially halted. No attempt to even tidy up intermediate allocations so cleanup won't work. Try running the tests with `JERRY_ENABLE_DEBUG=1` and see what happens!

A `reset()` function would be very easy to implement. However, depending on the use-case if we wanted multiple clients runninng together would we really want one of them to be able to pull the rug out from the others? Maybe, but that'll depend on usage.

**Pull request merged**